File: src/java/com/milo/protocols/thrift/TMiloProtocol.java
package com.milo.protocols.thrift;

import org.apache.thrift.protocol.*;
import org.apache.thrift.transport.TTransport;
import org.apache.thrift.TException;
import com.milo.storage.ThreadStorage;
import com.milo.tracer.TraceContext;
import com.milo.utils.SpanUtils;
import com.milo.protocols.thrift.TMiloMessageType;


public class TMiloProtocol extends TBinaryProtocol {
    private TraceContext traceContext = ThreadStorage.getTraceContext();

    public TMiloProtocol(TTransport tTransport) {
        super(tTransport);
    }

    public TMiloProtocol(TTransport tTransport, boolean b, boolean b1) {
        super(tTransport, b, b1);
    }

    /**
     * Factory
     */
    public static class Factory implements TProtocolFactory {
        protected boolean strictRead_ = false;
        protected boolean strictWrite_ = true;
        protected int readLength_;

        public TProtocol getProtocol(TTransport trans) {
            TMiloProtocol proto = new TMiloProtocol(trans, strictRead_, strictWrite_);
            if (readLength_ != 0) {
                proto.setReadLength(readLength_);
            }
            return proto;
        }
    }

    public void writeMessageBegin(TMessage message) throws TException {
        byte message_type = message.type;

        if (message.type == TMessageType.CALL && traceContext.isEnabled()) {
            message_type = TMiloMessageType.CALLTRACE;
        }
        writeString(message.name);
        writeByte(message_type);
        writeI32(message.seqid);

        if (message_type == TMiloMessageType.CALLTRACE) {
            writeI64(traceContext.getTrace_id());
            /* generate new 64-bit span id */
            writeI64(SpanUtils.generateID());
            writeI64(traceContext.getSpan_parent_id());
        }
    }

    public TMessage readMessageBegin() throws TException {
        int size = readI32();
        TMessage msg;

        if (size < 0) {
            int version = size & VERSION_MASK;
            if (version != VERSION_1) {
                throw new TProtocolException(TProtocolException.BAD_VERSION, "Bad version in readMessageBegin");
            }
            msg = new TMessage(readString(), (byte) (size & 0x000000ff), readI32());
        } else {
            if (strictRead_) {
                throw new TProtocolException(TProtocolException.BAD_VERSION, "Missing version in readMessageBegin, old client?");
            }
            msg = new TMessage(readStringBody(size), readByte(), readI32());
        }

        // set trace
        if (msg.type == TMiloMessageType.CALLTRACE) {
            long trace_id = readI64();
            int span_id = readI32();
            int span_parent_id = readI32();
            traceContext.setEnabled(true);
            traceContext.setRPCMethod(msg.name);
            traceContext.setTrace_id(trace_id);
            traceContext.setSpan_id(span_id);
            traceContext.setSpan_parent_id(span_parent_id);
        } else {
            traceContext.setEnabled(false);
        }

        return msg;
    }

}

File: web/thrift/lib/protocol/TMiloProtocol.php
<?php

include_once $GLOBALS['THRIFT_ROOT'].'/transport/TBufferedTransport.php';
include_once $GLOBALS['THRIFT_ROOT'].'/Thrift.php';
class TMilo_MessageType {
    const CALLTRACE = 5;
}

class TMiloProtocol extends TBinaryProtocolAccelerated {

  const VERSION_MASK = 0xffff0000;
  const VERSION_1 = 0x80010000;

  protected $strictRead_ = false;
  protected $strictWrite_ = true;

  public function __construct($trans, $trace_id=null) {
    parent::__construct($trans);
    $this->strictRead_ = false;
    $this->strictWrite_ = true;
    $this->trace_id = $trace_id;
  }

  public function writeMessageBegin($name, $type, $seqid) {
      if ($type == TMessageType::CALL && $this->trace_id !== null) {
        $type = TMilo_MessageType::CALLTRACE;
      }
      $version = self::VERSION_1 | $type;
      $msg = $this->writeI32($version) + $this->writeString($name) + $this->writeI32($seqid);

      if ($type == TMilo_MessageType::CALLTRACE) {
        list($span_id, $span_parent_id) = Milo_Util::incrementSpanID();
        $msg += $this->writeI64($this->trace_id) +
                $this->writeI32($span_id) +
                $this->writeI32($span_parent_id);
      }
      return $msg;
  }
}

/**
 * Accelerated milo binary protocol: used in conjunction with the thrift_protocol
 * extension for faster deserialization
 */
class TMiloProtocolAccelerated extends TMiloProtocol {
  public function __construct($trans, $strictRead=false, $strictWrite=true) {
    // If the transport doesn't implement putBack, wrap it in a
    // TBufferedTransport (which does)
    if (!method_exists($trans, 'putBack')) {
      $trans = new TBufferedTransport($trans);
    }
    parent::__construct($trans, $strictRead, $strictWrite);
  }
  public function isStrictRead() {
    return $this->strictRead_;
  }
  public function isStrictWrite() {
    return $this->strictWrite_;
  }
}

?>
